contrib/http_base/src/main/java/io/opentelemetry/contrib/http/core/AbstractHttpHandler.java

```diff
@@ -0,0 +1,204 @@
+import java.util.Map;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.annotation.Nullable;
+
+/** Base class for handling request on http client and server. */
+abstract class AbstractHttpHandler<Q, P> {
+
+  /** The {@link HttpExtractor} used to extract information from request/response. */
+  @VisibleForTesting final HttpExtractor<Q, P> extractor;
+
+  @VisibleForTesting final StatusCodeConverter statusConverter;
+  private final Meter meter;
+
+  /** Constructor to allow access from same package subclasses only. */
+  AbstractHttpHandler(
+      HttpExtractor<Q, P> extractor, StatusCodeConverter statusConverter, Meter meter) {
+    checkNotNull(extractor, "extractor is required");
+    this.extractor = extractor;
+    if (statusConverter == null) {
+      this.statusConverter = new StatusCodeConverter();
+    } else {
+      this.statusConverter = statusConverter;
+    }
+    if (meter == null) {
+      this.meter = OpenTelemetry.getMeterFactory().get(INSTRUMENTATION_LIB_ID);
+    } else {
+      this.meter = meter;
+    }
+  }
+
+  /**
+   * A convenience to record a {@link Event} with given parameters.
+   *
+   * @param span the span which this {@code Event} will be added to.
+   * @param type the type of event.
+   * @param uncompressedMessageSize size of the message before compressed (optional).
+   * @param compressedMessageSize size of the message after compressed (optional).
+   */
+  static void recordMessageEvent(
+      Span span,
+      long id,
+      AttributeValue type,
+      long uncompressedMessageSize,
+      long compressedMessageSize) {
+    Map<String, AttributeValue> attributes = new HashMap<>();
+    attributes.put(HttpTraceConstants.MSG_EVENT_ATTR_TYPE, type);
+    attributes.put(HttpTraceConstants.MSG_EVENT_ATTR_ID, AttributeValue.longAttributeValue(id));
+    if (compressedMessageSize > 0L) {
+      attributes.put(
+          HttpTraceConstants.MSG_EVENT_ATTR_COMPRESSED,
+          AttributeValue.longAttributeValue(compressedMessageSize));
+    }
+    if (uncompressedMessageSize > 0L) {
+      attributes.put(
+          HttpTraceConstants.MSG_EVENT_ATTR_UNCOMPRESSED,
+          AttributeValue.longAttributeValue(uncompressedMessageSize));
+    }
+    span.addEvent(HttpTraceConstants.MSG_EVENT_NAME, attributes);
+  }
+
+  private static void putAttributeIfNotEmptyOrNull(Span span, String key, @Nullable String value) {
+    if (value != null && !value.isEmpty()) {
+      span.setAttribute(key, AttributeValue.stringAttributeValue(value));
+    }
+  }
+
+  /**
+   * Instrument an HTTP span after a message is sent. Typically called for every chunk of request or
+   * response is sent.
+   *
+   * @param context request specific {@link HttpRequestContext}
+   * @param bytes bytes sent.
+   */
+  public final void handleMessageSent(HttpRequestContext context, long bytes) {
+    checkNotNull(context, "context is required");
+    context.sentMessageSize.addAndGet(bytes);
```

Comment:
mutating the context here seems very surprising. I would not expect that to be happening in a method with this name.

Reply:
This context is local to this handled. accumulating total bytes sent is needed to record total metric value for the entire span.